File: bst.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

struct Node {
    int val;
    Node* left;
    Node* right;
    Node(int v, Node* l = nullptr, Node* r = nullptr)
        : val(v), left(l), right(r) {}
};

Node* insert(Node* t, int v) {
    if (!t)          return new Node(v);
    if (v < t->val)  return new Node(t->val, insert(t->left,  v), t->right);
    if (v > t->val)  return new Node(t->val, t->left, insert(t->right, v));
    return t;
}

bool search(Node* t, int v) {
    if (!t)           return false;
    if (v == t->val)  return true;
    if (v < t->val)   return search(t->left,  v);
    return                   search(t->right, v);
}

void inorder(Node* t, std::vector<int>& out) {
    if (!t) return;
    inorder(t->left, out);
    out.push_back(t->val);
    inorder(t->right, out);
}

Node* build(const std::vector<int>& xs) {
    Node* t = nullptr;
    for (int v : xs) t = insert(t, v);
    return t;
}

double mean(const std::vector<int>& xs) {
    double s = 0;
    for (int x : xs) s += x;
    return s / xs.size();
}

double variance(const std::vector<int>& xs) {
    double m = mean(xs), s = 0;
    for (int x : xs) s += (x - m) * (x - m);
    return s / xs.size();
}

double stddev(const std::vector<int>& xs) { return std::sqrt(variance(xs)); }

double median(std::vector<int> xs) {
    std::sort(xs.begin(), xs.end());
    int n = xs.size();
    return (n % 2 == 0) ? (xs[n/2-1] + xs[n/2]) / 2.0 : xs[n/2];
}

int main() {
    std::vector<int> data = {64, 34, 25, 12, 22, 11, 90, 45, 67, 3};
    Node* tree = build(data);

    std::vector<int> result;
    inorder(tree, result);

    std::cout << "Sorted:    ";
    for (int x : result) std::cout << x << " ";
    std::cout << "\nMean:      " << mean(result)
              << "\nStd dev:   " << stddev(result)
              << "\nMedian:    " << median(result)
              << "\nSearch 45: " << std::boolalpha << search(tree, 45)
              << "\nSearch 99: " << std::boolalpha << search(tree, 99) << "\n";
    return 0;
}

File: bst.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val   = val
        self.left  = left
        self.right = right

def insert(tree, val):
    if tree is None:
        return Node(val)
    if val < tree.val:
        return Node(tree.val, insert(tree.left, val), tree.right)
    if val > tree.val:
        return Node(tree.val, tree.left, insert(tree.right, val))
    return tree

def search(tree, val):
    if tree is None:       return False
    if val == tree.val:    return True
    if val < tree.val:     return search(tree.left,  val)
    return                        search(tree.right, val)

def inorder(tree):
    if tree is None: return []
    return inorder(tree.left) + [tree.val] + inorder(tree.right)

def build(values):
    tree = None
    for v in values:
        tree = insert(tree, v)
    return tree

def mean(xs):
    return sum(xs) / len(xs)

def variance(xs):
    m = mean(xs)
    return sum((x - m) ** 2 for x in xs) / len(xs)

def std_dev(xs):
    return variance(xs) ** 0.5

def median(xs):
    s = sorted(xs)
    n = len(s)
    return (s[n//2-1] + s[n//2]) / 2 if n % 2 == 0 else s[n//2]

# ── main ──────────────────────────────────────────
data   = [64, 34, 25, 12, 22, 11, 90, 45, 67, 3]
tree   = build(data)
result = inorder(tree)

print(f"Sorted:    {result}")
print(f"Mean:      {mean(result):.2f}")
print(f"Std dev:   {std_dev(result):.2f}")
print(f"Median:    {median(result):.2f}")
print(f"Search 45: {search(tree, 45)}")
print(f"Search 99: {search(tree, 99)}")
